File: goldencheck/profilers/pattern_consistency.py
"""Pattern consistency profiler — detects inconsistent string patterns within a column."""
from __future__ import annotations
import polars as pl
from goldencheck.models.finding import Finding, Severity
from goldencheck.profilers.base import BaseProfiler

MINORITY_THRESHOLD = 0.30  # only flag patterns below this threshold
WARNING_THRESHOLD = 0.05  # <5% → WARNING (very rare, likely error); 5-30% → INFO


def _generalize(value: str) -> str:
    """Replace digits with D and letters with L, keeping punctuation as-is."""
    result = []
    for ch in value:
        if ch.isdigit():
            result.append("D")
        elif ch.isalpha():
            result.append("L")
        else:
            result.append(ch)
    return "".join(result)


class PatternConsistencyProfiler(BaseProfiler):
    def profile(self, df: pl.DataFrame, column: str, *, context: dict | None = None) -> list[Finding]:
        findings: list[Finding] = []
        col = df[column]

        if col.dtype not in (pl.Utf8, pl.String):
            return findings

        non_null = col.drop_nulls()
        total = len(non_null)
        if total == 0:
            return findings

        # Build pattern counts using Python (Polars map_elements for UDF)
        patterns = non_null.map_elements(_generalize, return_dtype=pl.String)
        pattern_counts = (
            patterns.value_counts()
            .sort("count", descending=True)
        )

        n_patterns = len(pattern_counts)
        if n_patterns <= 1:
            # All values share the same pattern — no inconsistency
            return findings

        dominant_count = pattern_counts["count"][0]
        dominant_pattern = pattern_counts[column][0]

        # Collect all minority patterns (rarest first — already sorted ascending by reversing)
        minority_candidates = []
        for i in range(1, n_patterns):
            minority_pattern = pattern_counts[column][i]
            minority_count = int(pattern_counts["count"][i])
            minority_pct = minority_count / total

            if minority_pct < MINORITY_THRESHOLD:
                minority_candidates.append((minority_pattern, minority_count, minority_pct))

        if not minority_candidates:
            return findings

        # Sort rarest first (most likely errors)
        minority_candidates.sort(key=lambda x: x[1])

        # Cap at top 5
        MAX_PATTERNS = 5
        total_minority = len(minority_candidates)
        emitted = minority_candidates[:MAX_PATTERNS]

        for minority_pattern, minority_count, minority_pct in emitted:
            # <5% → WARNING (very rare, likely error); 5-30% → INFO (valid variant)
            if minority_pct < WARNING_THRESHOLD:
                severity = Severity.WARNING
                confidence = 0.8
            else:
                severity = Severity.INFO
                confidence = 0.5
            # Find sample values that match this minority pattern
            mask = patterns == minority_pattern
            sample_vals = non_null.filter(mask).head(5).to_list()
            findings.append(Finding(
                severity=severity,
                column=column,
                check="pattern_consistency",
                message=(
                    f"Inconsistent pattern detected: '{minority_pattern}' appears in "
                    f"{minority_count} row(s) ({minority_pct:.1%}) vs dominant pattern "
                    f"'{dominant_pattern}' ({dominant_count} row(s))"
                ),
                affected_rows=minority_count,
                sample_values=[str(v) for v in sample_vals],
                suggestion="Standardize values to a single format/pattern",
                confidence=confidence,
            ))

        # Summary finding if more than MAX_PATTERNS minority patterns exist
        if total_minority > MAX_PATTERNS:
            extra = total_minority - MAX_PATTERNS
            findings.append(Finding(
                severity=Severity.INFO,
                column=column,
                check="pattern_consistency",
                message=(
                    f"{extra} additional inconsistent pattern(s) detected (showing top {MAX_PATTERNS})"
                ),
                suggestion="Standardize values to a single format/pattern",
                confidence=0.5,
            ))

        return findings

File: packages/goldencheck-js/src/core/profilers/pattern-consistency.ts
/**
 * Pattern consistency profiler — detects inconsistent string patterns.
 * Port of goldencheck/profilers/pattern_consistency.py.
 */

import type { TabularData } from "../data.js";
import { type Finding, Severity, makeFinding } from "../types.js";
import type { Profiler } from "./base.js";

const MINORITY_THRESHOLD = 0.30;
const WARNING_THRESHOLD = 0.05;
const MAX_PATTERNS = 5;

/** Replace digits with D and letters with L, keeping punctuation as-is. */
export function generalize(value: string): string {
  let result = "";
  for (const ch of value) {
    if (ch >= "0" && ch <= "9") result += "D";
    else if ((ch >= "a" && ch <= "z") || (ch >= "A" && ch <= "Z")) result += "L";
    else result += ch;
  }
  return result;
}

export class PatternConsistencyProfiler implements Profiler {
  profile(data: TabularData, column: string): Finding[] {
    const findings: Finding[] = [];

    if (!data.isString(column)) return findings;

    const nonNull = data.stringValues(column);
    const total = nonNull.length;
    if (total === 0) return findings;

    // Build pattern counts
    const patternCounts = new Map<string, number>();
    const patternValues = new Map<string, string[]>();

    for (const v of nonNull) {
      const p = generalize(v);
      patternCounts.set(p, (patternCounts.get(p) ?? 0) + 1);
      const samples = patternValues.get(p);
      if (!samples) {
        patternValues.set(p, [v]);
      } else if (samples.length < 5) {
        samples.push(v);
      }
    }

    if (patternCounts.size <= 1) return findings;

    // Sort by count descending
    const sorted = [...patternCounts.entries()].sort((a, b) => b[1] - a[1]);
    const dominantPattern = sorted[0]![0];
    const dominantCount = sorted[0]![1];

    // Collect minority patterns
    const minorities: Array<{ pattern: string; count: number; pct: number }> = [];
    for (let i = 1; i < sorted.length; i++) {
      const [pattern, count] = sorted[i]!;
      const pct = count / total;
      if (pct < MINORITY_THRESHOLD) {
        minorities.push({ pattern, count, pct });
      }
    }

    if (minorities.length === 0) return findings;

    // Sort rarest first
    minorities.sort((a, b) => a.count - b.count);

    const emitted = minorities.slice(0, MAX_PATTERNS);
    for (const { pattern, count, pct } of emitted) {
      const severity = pct < WARNING_THRESHOLD ? Severity.WARNING : Severity.INFO;
      const confidence = pct < WARNING_THRESHOLD ? 0.8 : 0.5;
      const sampleVals = patternValues.get(pattern)?.slice(0, 5) ?? [];

      // Detect structural pattern shift (letter-first vs digit-first)
      const domStartsAlpha = dominantPattern.length > 0 && dominantPattern[0] === "L";
      const minStartsAlpha = pattern.length > 0 && pattern[0] === "L";
      const msgExtra =
        domStartsAlpha !== minStartsAlpha && pct < WARNING_THRESHOLD
          ? " — possible invalid format or mixed coding standard"
          : "";

      findings.push(
        makeFinding({
          severity,
          column,
          check: "pattern_consistency",
          message:
            `Inconsistent pattern detected: '${pattern}' appears in ` +
            `${count} row(s) (${(pct * 100).toFixed(1)}%) vs dominant pattern ` +
            `'${dominantPattern}' (${dominantCount} row(s))` +
            msgExtra,
          affectedRows: count,
          sampleValues: sampleVals,
          suggestion: "Standardize values to a single format/pattern",
          confidence,
          metadata: { dominant_pattern: dominantPattern, minority_pattern: pattern },
        }),
      );
    }

    // Summary if more patterns exist
    if (minorities.length > MAX_PATTERNS) {
      const extra = minorities.length - MAX_PATTERNS;
      findings.push(
        makeFinding({
          severity: Severity.INFO,
          column,
          check: "pattern_consistency",
          message: `${extra} additional inconsistent pattern(s) detected (showing top ${MAX_PATTERNS})`,
          suggestion: "Standardize values to a single format/pattern",
          confidence: 0.5,
        }),
      );
    }

    return findings;
  }
}
